### PR TITLE
Modbus binding for openhab 2

### PR DESCRIPTION
**Pull request by vores8**
Tested with tcp and serial comms.

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Thanks, I will try to review in the coming weeks and come back to you!

**Comment by jensriebold** [comment]
I would like to use the openhab2 modbus bridge for my project. What are the problems why this pull is not integrated into the openhab2 master?

**Comment by jensriebold** [comment]
Hello,

I just merged the PULL request in my GIT repository and installed openHAB 2 on my raspberry pi.
The RXTX lib is located in the /usr/share/java directory.

How can I set the CLASSPATH that java 8 will use the RXTXcomm-2.2pre2.jar?

Since the RXTXcomm depends on native code it make sense that the debian version is used.

Here the exception ( It is pretty clear what happend )

Jens Riebold

00:04:36.563 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/opt/openhab2/addons/org.openhab.binding.modbus-2.0.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.modbus [168]
  Unresolved requirement: Import-Package: gnu.io

        at org.eclipse.osgi.container.Module.start(Module.java:434)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:393)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1245)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1217)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:509)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:358)[8:org.apache.felix.fileinstall:3.5.0]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:310)[8:org.apache.felix.fileinstall:3.5.0]

**Comment by vores8** [comment]
You need gnu.io library, it replaces rxtx.

2016-01-20 3:19 GMT+03:00 Jens Riebold <notifications@github.com>:

**Comment by kaikreuzer** [comment]
Please simply run
```
feature:install openhab-transport-serial
```
in the openHAB shell. His will install nrjavaserial, which packages rxtx (which implements gnu.io) and automatically unpacks the suitable binaries for your platform.

**Comment by jensriebold** [comment]
Thanks a lot for the hint.

It works after I removed the the openhab1 modbus binding from 
/opt/openhab2/runtime/karaf/system/org/openhab/addons/openhab-addons/1.9.0-SNAPSHOT

The next step is now to implement a binding for my homemade solar controller that uses the modbus RTU protocol.

So I use the this modbus binding as a bridge to implement bindings for all my homemade modbus devices.

Am I on the right path?

**Comment by jensriebold** [comment]
Hello,

I got a NPE when I installed the modbus binding:

/org.openhab.binding.modbus/ESH-INF/config/config.xml

  <config-description uri="thing-type:modbus:config">
  
  </config-description>

I added the URI uri="thing-type:modbus:config"

Then the binding was successfully started. I think that here are missing configuration parameter like modbus address and a reference to the binding that uses the bridge?

**Comment by vores8** [comment]
The common syntax is

Bridge modbus:tcp:conn1 {
Bridge slave reg1 [length=2, type="holding"] {
Thing endpoint sw1 [read=0, write=1]
}

1st string defines connection type. Can be
Bridge modbus:tcp:connection_name [ip=network_address, port=network_port] {
... }
ip defaults to localhost, port defaults to 502
Or it can be
Bridge modbus:serial:connection_name [port= , baud=, databits= , stopbits=
, parity= , encoding=] {...}
default baud = 9600
default databits = 8 (can be 5 6 7 8)
default stopbits = 1 (can be 1 1.5 2)
default parity = none (none even odd)
default encoding = ascii (ascii rtu bin)

2nd line defines modbus slave, parameters pretty much from v1 binding
Bridge slave slave_name [refresh= , id= , start=, length= , type= ] {
refresh default to 500 (value in msec)
id defaults to 1
start defaults to 0
type can be coil discrete holding input

3rd line defines endpoint - actual modbus register or register pair again
like in v1
Thing endpoint item_name [read= , write= ]
read defaults to 0
write is optional
item types supported: switch contact number

You can specify multiple slaves inside a single connection
You can specify multiple endpoints inside a single slave






2016-01-21 3:14 GMT+03:00 Jens Riebold <notifications@github.com>:

**Comment by jensriebold** [comment]
Thanks, your modbus bundle is started correctly, I can configure the binding via paperUI.
It quite hard get everything running, since I just started with openHAB2 (with no experience on OpenHAB1)

**Comment by jensriebold** [comment]
The modbus bridge configures things by direct association of a modbus transaction. This is perfect for monitoring the device.

For the configuration I need direct access to the modbus bridge, because the configuration of my solarcontroller is held in the eeprom of the device. So the binding has to perform some modbus sequences on startup. 

How can I associate the Bridge to my binding?

**Comment by vores8** [comment]
Sorry I don't understand this. It's just a modbus binding that implements
pure modbus protocol. Modbus protocol does not implement any device-centric
behaviour. If you need custom interaction  then you need to create your own
device-specific binding on top of modbus protocol implementation

2016-01-23 22:08 GMT+03:00 Jens Riebold <notifications@github.com>:

**Comment by jensriebold** [comment]
Some pseudo code:
jrsolarcontroller --> this is my binding

Bridge modbus:serial:serhome [port=/dev/ttyUSB0 , baud=38400, databits=8 , stopbits=1 , parity=even , encoding=rtu] 
{ Bridge slave solarcontroller [id=10, binding=jrsolarcontroller]}

And in my binding I use your pure modbus binding to communicate with the device.

**Comment by jensriebold** [comment]
Hello,

I think the best solution to make the modbus binding more useable is to build an OSGi bundle that covers only the pure modbus protocol.

For every modbus master an instance is created with an own queue that serialize the requests.

The modbus binding tracks the modbus bundle and deligates to the particular instance.
Then other modbus implementations also work with the modbus bundle.

What about this idea?

**Comment by watou** [comment]
That sounds like the right idea!

**Comment by ssalonen** [comment]
@vores8 

I wonder would it make sense to check whether any changes and fixes introduced in PR #4226 (for openhab1) would apply here?

**Comment by Jochen1980** [comment]
Hi! Are there any updates on this binding or the mentioned approaches?

File: addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/SlaveHandler.java [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+            } else if (DATA_TYPE_DISCRETE.equalsIgnoreCase(sType)) {
+                type = DATA_TYPES.DISCRETE;
+            } else if (DATA_TYPE_HOLDING.equalsIgnoreCase(sType)) {
+                type = DATA_TYPES.HOLDING;
+            } else if (DATA_TYPE_INPUT.equalsIgnoreCase(sType)) {
+                type = DATA_TYPES.INPUT;
+            }
+
+        } catch (Exception e) {
+        }
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/SlaveHandler.java]: log something here?

File: addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/SlaveHandler.java [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        return r;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        for (Thing t : this.getBridge().getThings()) {
+            EndpointHandler h = (EndpointHandler) t.getHandler();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/SlaveHandler.java]: This method is doing NOTHING. Is this on purpose?

**Comment by kaikreuzer** [comment]
@vores8 I have done a review and created https://github.com/vores8/openhab2/pull/1 with some updates, which you can simply merge so that it is included in this PR.

Imho there are two points that ask for a refactoring:

- As discussed above, it would make sense to introduce an org.openhab.io.transport.modbus bundle, which provides low-level modbus communication and that can be used from different bindings. Most code of that new bundle is probably anyhow from the jamod library.
- We should not include the full sources from other projects that are under a different license. I have therefore removed the sources and replaced them by [the official download](https://sourceforge.net/projects/jamod/files/jamod/1.2/jamod-1.2-SNAPSHOT.jar/download). I understand that changes to this lib were done in order to switch from javax.comm to gnu.io. Nonetheless, these changes should be done outside of the binding - I am fine with creating a fork of the project at https://github.com/openhab/jamod, where this can be maintained. The resulting library should then be added to the binding (or rather the io.transport.modbus bundle).

Would you all agree with this procedure?

**Comment by vores8** [comment]
Kai thanks Unfortunately I am a bit busy these days, so please do not expect any extensive contribution. Sorry about that ThanksDmitry20.07.2016, 11:31, "Kai Kreuzer" <notifications@github.com>:@vores8 I have done a review and created vores8#1 with some updates, which you can simply merge so that it is included in this PR.

Imho there are two points that ask for a refactoring:


As discussed above, it would make sense to introduce an org.openhab.io.transport.modbus bundle, which provides low-level modbus communication and that can be used from different bindings. Most code of that new bundle is probably anyhow from the jamod library.
We should not include the full sources from other projects that are under a different license. I have therefore removed the sources and replaced them by the official download. I understand that changes to this lib were done in order to switch from javax.comm to gnu.io. Nonetheless, these changes should be done outside of the binding - I am fine with creating a fork of the project at https://github.com/openhab/jamod, where this can be maintained. The resulting library should then be added to the binding (or rather the io.transport.modbus bundle).


Would you all agree with this procedure?

**Comment by ssalonen** [comment]
@kaikreuzer I like the idea in the sense it might also help separating openhab1 binding code from openhab2. Would like to approach where both binding code can be maintained easily. As you probably know, there are also instances where the modbus binding has been forked to support some protocol (could not find the one I had in mind now). WAMO binding seems to [use plain jamod](https://github.com/openhab/openhab/tree/master/bundles/binding/org.openhab.binding.wago/lib). Having low level bundle would certainly help.

- Do you propose that the forked jamod would be git submodule of io.transport.modbus bundle? In total we would have three libraries: jamod fork (no OSGI code at all), low level bundle, high level bundle
- In the latest [PR #4226](https://github.com/openhab/openhab/pull/4226) from me connections are shared such that only single connection can be active to one end point. Do you agree that this fits to the io.transport.modbus bundle? By having it there, all the bindings using modbus would have the same connection limits ([communication "serialized"](https://github.com/openhab/openhab2-addons/pull/362/#issuecomment-177055094)). This separation naturally requires some effort, not sure how OSGI fits to all of this...
- Would the high level bundle include configuration handling (creating slaves based on configuration), handling commands and setting up polling but not much else?

**Comment by kaikreuzer** [comment]
Hi @ssalonen (sorry for the late reply, I was on vacation...),

> Do you propose that the forked jamod would be git submodule of io.transport.modbus bundle?

No, not really - it could be built separately and then simply included as a jar in the io.transport.modbus bundle.

> Do you agree that this fits to the io.transport.modbus bundle?

I guess so - the io.transport.modbus would be the "connection broker" to modbus and manages the access for all bundles that want to make use of it. I am not in the details on what this means for modbus, but it seems you are a real expert there, so you will be able to tell!

> Would the high level bundle include configuration handling (creating slaves based on configuration), handling commands and setting up polling but not much else?

Again, I am not in the details on what are the normal use cases and requirements of people here. But yes, I would think that a "generic" modbus binding would offer such stuff. Additionally, there might then also be "specific" modbus-bindings that provide support for a specific device that speaks modbus.

**Comment by ssalonen** [comment]
Thanks for the answers...I think I start to see the point here. Couple of questions still :
- Do we have an example of this kind of separation of libraries? I can see that we use osgi "require bundle", but not really copying jars. I'm thinking if this could be automated with maven as part of build process (since we are actually touching jamod code as well), what do you think? What's the reason for not going the "require bundle way"?

- I think broker APIs definitely need work in order to separate the actual protocol and connection  handling from the openhab events (interpretation of register data, and sending events). Do we have any use cases in mind? I now found the example of forking the modbus binding, see [this thread](https://community.openhab.org/t/helios-binding-for-modbus-tcp/5578/7). Having separate low level transport bundle would help here, I think.

**Comment by lewie** [comment]
Any news here? Is this same as here? https://github.com/openhab/openhab2-addons/issues/107

**Comment by ssalonen** [comment]
#107 is new to me. Have to check it out.

I'm personally working on the binding for openHAB2. I managed to create new low level api in transport for other bindings to use. As a exercise and for verification purposes (there are quite many integration tests) I ported the existing openHAB1 binding to use this API. 

The result is on my github https://github.com/ssalonen/openhab/tree/modbus-openhab2?files=1. Note that in that branch, all the code still lives in the same bundle. I have started the physical bundle separation here https://github.com/ssalonen/openhab/tree/modbus-openhab2-transport-binding-separation?files=1. That is still work in progress.

So far all the imports are openhab1, naturally it would make sense to transform those smarthome equivalents, but I prefer to do this in steps and take adventure of the vast integration tests there.

Have not yet started to work on the actual openhab2 binding but I do have some ideas about it. I plan to drop support for some stuff (e.g. Handling of increase commands) and clean something else (32bit value types would be available as readonly channels, since there is not yet write implementation for those).

The transport API makes sure that connections are re-used and not too many transactions are initated to the same slave device. These were some of the main issues the binding were suffering before the 1.9.0 connection pooling.

Transport api provides methods for one-time poll, regular poll, and one-time write coil / holding register writes.

What should be best way coordinate the work? New "wip" pull request to pool the discussions or...? I would like to involve the community forums at a later stage when there is something to actual test out.

Best 
Sami

**Comment by dexter1515** [comment]
There is a successor of the jamod library called j2mod under active development.
https://github.com/steveohara/j2mod
Maybe this could be an alternative to including the full source of jamod like @kaikreuzer mentioned before.

**Comment by ssalonen** [comment]
I am aware of this fork, note that the link you pasted is actually fork of the original j2mod.

I think I have listed my observatorions on the different forks here in the community forums https://community.openhab.org/t/communication-with-unvelvent-vmc-domeo-450-with-modbus/17185/10?u=ssalonen. Some concerns were related to support of many cpu architectures.

Please also note that openhab jamod in modbus is fork of original jamod, see the link above listing some of the changes done there.

Definitely think we can consider alternative libraries although I would postpone that as a separate step when we have stable 2.0 binding (at least as stable as pre 2.0 openhab). What do you think?

Best 
Sami

**Comment by ssalonen** [comment]
To keep everyone up to date here: started a discussion on thing hiearchy on the community forums: https://community.openhab.org/t/seeking-feeedback-for-modbus-openhab2-native-binding/27657 . I'm trying to keep the features developed in openHAB1 version in the new binding as well.

**Comment by kaikreuzer** [comment]
Thanks for pushing this topic forward @ssalonen and sorry that I dropped the ball on it and never responded to your last questions - I guess you have figured out the answers yourself by now (if not, try again to ask, please)...

Since https://github.com/vores8/openhab2/pull/1 isn't merged or commented since almost a year now, I assume that @vores8 is not actively developing it anymore. So maybe it would be best to close this PR and create a new [WIP] one from @ssalonen's branch as a follow up. Wdyt?

**Comment by ssalonen** [comment]
@kaikreuzer Let's do that.

PR #2246 created.

**Comment by kaikreuzer** [comment]
Thanks!

**Pull request closed without merging**